teps/0005-tekton-oci-bundles.md

```diff
@@ -0,0 +1,481 @@
+./pipelines/
+  /go-release
+    /0.1
+      /README.md
+      /go-release.yaml
+      /samples/dummy-go-release.yaml
+```
+
+The catalog infrastructure could automate building and publishing
+`Task`s and `Pipeline`s based on this organization. For the above
+example, you would have the following (assuming the image reference
+prefix would be `gcr.io/tekton-catalog/…`):
+
+```
+gcr.io/tekton-catalog/task/argocd:0.1
+gcr.io/tekton-catalog/task/argocd:0.2
+gcr.io/tekton-catalog/task/golang-build:0.1
+gcr.io/tekton-catalog/pipeline/go-release:0.1
+```
+
+We could even make sure the Tekton OCI bundle that include the
+`Pipeline` definition would also include the `Task`s definition it
+requires. For the `go-release` pipeline, this would mean it would
+include a layer with the `Pipeline` definition, a layer with the
+`go-build` `Task`, etc.
+
+#### Tooling
+
+To give users a default way of generating Tekton Bundles, we could add
+a set of commands to `tkn` to support building, inspecting, and
+modifying images.
+
+The logic behind the commands below is that each image can and will
+probably contain multiple different types of objects so it doesn’t fit
+the existing scheme of `tkn`. Instead we choose a new top level “remote”
+action to act on the collective set of tekton objects remotely whether
+that be images, or possibly in a catalog store in github in the
+future.
+
+```shell
+# Generates an image. Can be passed files, directories, etc. and will add each found Tekton object to the image. Supports tagging the image by a name
+tkn remote build [REF]
+
+# Pushes the image up to a remote registry
+tkn remote push [REF]
+
+# Fetches all of the contents of an image and prints them in a terse format. Looks for the local copy before looking for a remote copy.
+tkn remote ls [REF]
+# task / my-task
+# pipeline / my-pipeline
+
+# Dumps the contents in a flat directory.
+tkn remote get [REF]
+
+# Returns the specified object.
+tkn remote get [REF] [KIND] [NAME]
+
+# These work like the existing commands but from an image, not a namespace.
+tkn task list --image=[REF]
+tkn task get --image=[REF] [NAME]
+# These apply to the local "build" of the image. If it doesn't exist, it is created from scratch
+tkn task create --image=[REF] -f foo.yml
+tkn task delete --image=[REF] -f foo.yml
+
+tkn task start --image=[ref] [NAME]
+# ... all the other resources get this as well
+```
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+- Using OCI images and layer for what they are not intended for — at
+least with the usual `MIME` type — might be confusing for users. A
+`docker pull gcr.io/tekton-catalog/task/golang-build:0.1` will fail
```

Comment:
At a minimum, we should document the error message you'll see when you `docker pull` it, so that users can find it and learn what it means (e.g., "You tried to pull a Tekton OCI bundle, which is not runnable").

Going further, do we have any say in _how_ it fails? For instance, we could document/require (and support with tooling) that the image should actually have a valid MIME type, and when it's `docker run`, it will print some useful error message? Do helm images do anything like this?